Fix --start-at-task

s-hertel opened the pull request:
##### SUMMARY

Fixes #86268

~Marked as a draft because this needs some more discussion.~  Concerns resolved, I think.

##### ISSUE TYPE

- Bugfix Pull Request

s-hertel committed:
Fix fact gathering with --start-at-task

Add a VALIDATE state between SETUP and TASKS

Remove now unused pending_setup attribute

ci_complete

mkrizek commented on lib/ansible/executor/play_iterator.py:
```diff
@@ -296,28 +297,26 @@ def _get_next_task_from_state(self, state, host):
-                    elif not gather_facts and not self._play.validate_argspec:
+                elif state.pending_setup:
+                    task = self._blocks[0].block[1]
+                    if not self._play.validate_argspec:
```
I wonder if creating a new state (`IteratingStates.ValidatePlayArgspec` or something) would simplify the implementation?

webknjaz commented on lib/ansible/executor/play_iterator.py:
```diff
@@ -296,28 +297,26 @@ def _get_next_task_from_state(self, state, host):
                         task = None
-                    elif not gather_facts and not self._play.validate_argspec:
+                elif state.pending_setup:
+                    task = self._blocks[0].block[1]
```
```
14:14     task = self._blocks[0].block[1]
14:14            ~~~~~~~~~~~~~~~~~~~~~^^^
14:14 IndexError: list index out of range
```

s-hertel committed:
Update changelog to be more general

ci_complete

mkrizek commented on lib/ansible/executor/play_iterator.py:
```diff
@@ -69,7 +70,6 @@ def __init__(self, blocks):
         self.fail_state = FailedStates.NONE
         self.pre_flushing_run_state = None
         self.update_handlers = True
-        self.pending_setup = False
```
🎉

mkrizek commented on lib/ansible/executor/play_iterator.py:
```diff
@@ -37,11 +37,12 @@
-    ALWAYS = 3
-    HANDLERS = 4
-    COMPLETE = 5
+    VALIDATE = 1
```
I see why we'd want to preserve the order but strictly speaking this is a breaking change. I think it's fine to add `VALIDATE = 6`.

s-hertel replied:
The Python API and strategy plugins have no guarantees about backward compatibility, but I've updated it as you suggest and hardcoded the gather_facts and validate_argspec indices instead.

mkrizek replied:
I guess I didn't realize the deprecation of 3rd party strategy plugins was with an immediate effect.

s-hertel replied:
I think even just adding a new state is not really backward compatible. Any 3rd party strategy plugins that were running `SETUP` tasks will not necessarily handle the new `VALIDATE`. However, after poking around on Github I don't see anything obvious and current that would be broken (it actually looks like some projects have removed their custom strategies that would have been impacted by this change, like https://github.com/openstack-k8s-operators/edpm-ansible/commit/41c7a3d134f5026428300383848278bcf5fd4f9b).

I didn't think many 3rd party strategy plugins would be impacted by adding a new state or changing the order since many don't inherit from the base class or use these classes directly, and instead just piggyback on specific builtin strategies.

mkrizek commented on lib/ansible/executor/play_iterator.py:
```diff
@@ -292,51 +290,37 @@ def _get_next_task_from_state(self, state, host):
+                    (gathering == 'explicit' and boolean(self._play.gather_facts, strict=False)) or
+                    (gathering == 'smart' and implied and not self._variable_manager._facts_gathered_for_host(host.name))
+                ):
+                    task = self._blocks[0].block[state.run_state]
```
I think we are overloading the use of `run_state` here. We already have a way to point into the current block by using `cur_regular_task`:

```python
if gather_facts ...:
    task = block.block[state.cur_regular_task]
state.run_state = IteratingStates.VALIDATE
state.cur_regular_task += 1  # point at the validate task, for the next state
```

mkrizek replied:
Of course after submitting I realized why you used `run_state` :-) I suppose we can either hardcode the indexes for setup and validate (`0`, `1`) or (potentially for a separate PR) I wonder why do we even use blocks when we can just store the setup and validate tasks on `PlayIterator` and returned them when needed? Just an idea.

s-hertel replied:
Yeah, `run_state` is what `--start-at-task` uses to execute the play-level tasks before starting at the requested task. `cur_regular_task` is saving the task to start at, and should not be modified here.

I've updated it to use hardcoded indices.

> ... I wonder why do we even use blocks when we can just store the setup and validate tasks on PlayIterator and returned them when needed? Just an idea.

The block is used to filter the tasks by tag, so it would need some refactoring. I'm not sure if it's worth it.

mkrizek commented on lib/ansible/executor/play_iterator.py:
```diff
@@ -292,51 +290,37 @@ def _get_next_task_from_state(self, state, host):
+                # done it for this host; or if 'explicit' and the play sets
+                # gather_facts to True; or if 'implicit' and the play does
+                # NOT explicitly set gather_facts to False.
+                gather_facts = len(self._blocks[0].block) > state.run_state
```
Do we need this check? We seem to always put the `setup_task` into the first block. Same for the new `VALIDATE` state.

s-hertel replied:
The tasks in the block are filtered by tags, so assuming self._blocks[0].block[0] exists can cause an IndexError.

s-hertel replied:
> Same for the new VALIDATE state.

Found a bug thanks to that. I even wrote the test case wrong, so I fixed that.

mkrizek commented on lib/ansible/executor/play_iterator.py:
```diff
@@ -292,51 +290,37 @@ def _get_next_task_from_state(self, state, host):
-                # clear the pending setup flag, since we're past that and it didn't fail
-                if state.pending_setup:
-                    state.pending_setup = False
+            elif state.run_state == IteratingStates.VALIDATE:
```
I think we need to add `VALIDATE` handling to `_set_failed_state`; should be just a copy-paste of the `SETUP` state there.

s-hertel replied:
Good catch, added.

mkrizek reviewed (changes requested):
Thanks for splitting the validation into a separate state. Just one concern and a few suggestions.

s-hertel committed:
Add new states for IteratingStates and FailedStates without changing existing states

Fix a bug with validate_argspec being tagged with always even it will inherit play-level tags. Now it aligns with the play-level gather_facts task's use of tags.

ci_complete
